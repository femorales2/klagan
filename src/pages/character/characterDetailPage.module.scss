.container {
  display: flex;
  flex-direction: column;
  gap: 54px;
  align-items: center;
  padding-bottom: 30px;
}