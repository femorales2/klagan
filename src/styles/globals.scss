:root {
  --font-mono: ui-monospace, Menlo, Monaco, 'Cascadia Mono', 'Segoe UI Mono',
    'Roboto Mono', 'Oxygen Mono', 'Ubuntu Monospace', 'Source Code Pro',
    'Fira Mono', 'Droid Sans Mono', 'Courier New', monospace;

  --navbar-bg: #000;
  --secondary-color: #fff;
  --marvel-red: #ec1d24;
  --scroll-track-bg: #d9d9d9;
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

a {
  display: flex;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}
