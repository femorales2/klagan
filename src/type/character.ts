import { PUBLIC_KEY } from '@/constants/api';

export type Character = {
  id: number | string;
  name: string;
  description: string;
  thumbnail: Thumbnail;
  comics: {
    available: number;
    items: Comic[];
  }
}

export type Comic = {
  name: string;
  resourceURI: string;
}

export type CharacterResponse = {
  data: {
    results: Character[];
    count?: number;
  }
}

export type Thumbnail = {
  path: string;
  extension: string;
}

export type ParamsType = {
  apikey: string;
  limit: string;
  offset: string;
  hash?: string;
  ts?: string;
}