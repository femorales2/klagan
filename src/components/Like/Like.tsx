import React, { useMemo, MouseEvent } from 'react';
import { useCardContext } from '@/context/CardContext';
import EmptyHeart from '@/components/EmptyHeart/EmptyHeart';
import Heart from '@/components/Heart/Heart';

interface ILike {
  characterId: number | string;
  size?: number;
}

const Like = ({characterId, size = 10}: ILike) => {
  const {toggleFavorite, favorites} = useCardContext();
  
  const isFavorite = useMemo(() => favorites.some(favorite => favorite.id === characterId), [favorites, characterId]);
  
  const handleClick = (e: MouseEvent) => {
    e.stopPropagation();
    toggleFavorite({id: characterId});
  }
  
  return (
    <span onClick={handleClick}>
      {isFavorite ? <Heart width={size} height={size} /> : <EmptyHeart width={size} height={size} />}
    </span>
  );
};

export default Like;