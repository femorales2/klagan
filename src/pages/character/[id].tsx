import React, { useEffect, useLayoutEffect } from 'react';
import { GetServerSideProps } from 'next';
import CharacterDetail from '@/components/CharacterDetail/CharacterDetail';
import styles from "./characterDetailPage.module.scss";
import CharacterComics from '@/components/CharacterComics/CharacterComics';
import { useCardContext } from '@/context/CardContext';

interface ICharacterDetailPage {
  [index: string]: string;
  id: string;
}

const Page = ({id}: ICharacterDetailPage) => {
  
  const {setSelectedId} = useCardContext();
  
  useLayoutEffect(() => {
    console.log("page id hook", id);
    setSelectedId(id);
  }, [id, setSelectedId]);
  
  return (
    <section className={styles.container}>
      <CharacterDetail />
      <CharacterComics />
    </section>
  );
};

export const getServerSideProps: GetServerSideProps<ICharacterDetailPage, ICharacterDetailPage> = async (context) => {
  const id = context.params?.id;
  /**
   * if there isn't id nor id is a number
   * we shall redirect to home
   */
  if (!id || isNaN(parseInt(id))) {
    return {
      redirect: {
        destination: '/',
        permanent: false
      }
    };
  }
  
  return {props: {id}};
}

export default Page;