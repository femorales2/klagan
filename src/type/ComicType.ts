import { Thumbnail } from '@/type/character';

export type ComicType = {
  id: number;
  title: string;
  thumbnail: Thumbnail;
  dates: ComicDate[];
}

export type ComicDate = {
  date: string;
  type: string;
}

export type ComicResponse = {
  data: {
    results: ComicType[];
  }
}