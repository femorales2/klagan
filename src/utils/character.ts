import { Character } from '@/type/character';

export const getFavoritesIds = (favorites: Character[]) => {
  
  return favorites.map(favorite => favorite.id);
  
}