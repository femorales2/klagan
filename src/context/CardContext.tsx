import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useEffect,
  useState
} from 'react';
import useRequester from '@/hooks/useRequester';
import { fetchCharacters } from '@/utils/api';
import { Character, CharacterResponse } from '@/type/character';

export interface ICardContext {
  data: CharacterResponse['data'];
  search: string;
  setSearch: Dispatch<SetStateAction<string>>;
  favorites: Character[];
  selectedId: string;
  toggleFavorite(character: Partial<Character>): void;
  setSelectedId: Dispatch<SetStateAction<string>>;
  enabledFavoriteView: boolean;
  setEnableFavoriteView: Dispatch<SetStateAction<boolean>>;
}

export const CardContext = createContext<ICardContext>({
  favorites: []
} as unknown as ICardContext);

export const useCardContext = () => useContext(CardContext);

const CardContextProvider = ({ children }: { children: React.ReactNode }) => {
  const [cardData, setCardData] = useState<ICardContext['data']>({
    results: []
  });
  const [limit, setLimit] = useState<number>(50);
  const [search, setSearch] = useState<string>('');
  const [selectedId, setSelectedId] = useState<string>('');
  const [favorites, setFavorites] = useState<Character[]>([]);
  const [enabledFavoriteView, setEnableFavoriteView] = useState<boolean>(false);
  const { result, loading } = useRequester<CharacterResponse>({
    requestCallback: () => fetchCharacters({ limit })
  });

  const toggleFavorite = (character: Character) => {
    const exists = favorites.some((favorite) => favorite.id === character.id);
    if (exists) {
      setFavorites(
        favorites.filter((favorite) => favorite.id !== character.id)
      );
    } else {
      setFavorites([...favorites, character]);
    }
  };

  useEffect(() => {
    if (!loading && result?.data.results.length) {
      setCardData(result.data);
    }
  }, [loading, result]);

  return (
    <CardContext.Provider
      value={{
        data: cardData,
        search,
        setSearch,
        favorites,
        toggleFavorite,
        selectedId,
        setSelectedId,
        enabledFavoriteView,
        setEnableFavoriteView
      }}
    >
      {children}
    </CardContext.Provider>
  );
};

export default CardContextProvider;
