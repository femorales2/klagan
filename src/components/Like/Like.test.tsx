import { describe, expect, it, jest } from '@jest/globals';
import { render, fireEvent, getByTestId } from '@testing-library/react';
import Like from '@/components/Like/Like';
import { CardContext, ICardContext } from '@/context/CardContext';
import { Character } from '@/type/character';

describe('<Like />', () => {
  it('should render', () => {
    const {container} = render(
      <CardContext.Provider value={{
        favorites: [
          {
            id: 1
          }
        ],
      } as ICardContext}>
        <Like characterId={1} />
      </CardContext.Provider>
    );
    
    expect(container).toMatchSnapshot();
  });
  
  it('should call toggleFavorite', () => {
    const toggleFavorite = jest.fn();
    const {container} = render(
      <CardContext.Provider value={{
        favorites: [
          {
            id: 1
          }
        ],
        toggleFavorite: toggleFavorite as (char: Character) => void
      } as ICardContext}>
        <Like characterId={1} />
      </CardContext.Provider>
    );
    
    fireEvent.click(getByTestId(container, "like"));
    
    expect(toggleFavorite).toBeCalled();
  });
});