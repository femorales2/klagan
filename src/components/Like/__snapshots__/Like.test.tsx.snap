// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Like /> should render 1`] = `
<div>
  <span>
    <svg
      fill="none"
      height="10"
      viewBox="0 0 13 12"
      width="10"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M6.57153 2.37318L3.57153 0.552368L0.571533 2.37318V6.27491L6.57153 11.3905L12.5715 6.27491V2.37318L9.57154 0.552368L6.57153 2.37318Z"
        fill="#EC1D24"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</div>
`;
